### PR TITLE
CTL-1308: EventArgsConverter Support in EventToCommand

### LINKED ISSUE #1308: [Feature]: EventArgsConverter in EventToCommand
### Preamble

When developing an application, it is usually inevitable that a View type of event has to be handled and a routine within the ViewModel has to be called. Depending on this event there are generally two pathways in which this handled:

**Direct Implementation**
Event is implemented in the View code-behind and the ViewModel is not required as part of the interaction (a service might be though)

**Mixed-Mode Implementation**
Event is implemented in the View code-behind and the ViewModel is required. Implementations such as this come in two types, unidirectional and bidirectional. Unidirectional implementations are for when you need short-circuit the event based on event conditions before passing a call to the ViewModel and do not want to leak the event types to the ViewModel. Bidirectional implementations are where you need to manipulate properties within an Event using results from a method executed in the ViewModel. An example of this would be [DistinctValuesLoading Event](https://docs.telerik.com/devtools/wpf/controls/radgridview/filtering/how-to/howto-display-all-distinct-values) from Telerik's RadGridView.

**EventToCommand Converter**
Event is implemented in the ViewModel as the View is not directly dependent on it. A primitive example would be any **Loaded** event for a **UIFrameworkElement** type object. However there is an inherit downside with this implementation if one must use the **PassEventArgsToCommand** dependency property if for whatever reason **CommandParameter** cannot be used against a require set of object parameters due to limitations within the UI framework of choice.

The downside is that View-level events get exposed into the ViewModel which means the ViewModel is no longer agnostic to the View implementation. That's where this feature idea comes in.

### Feature description

The premise of this is feature is quite straight forward hopefully, the concept is a Dependency Property called **EventArgsConverter** would be introduced within **EventToCommand** that allows an implementer to introduce basically an object bridge between the View and ViewModel that would convert any **EventArgs** implementation into a neutral object that the ViewModel can assimilate. 

There are some limitations to this feature however, while it can help eliminate the need to implement code for a Mixed-Mode Unidirectional implementation and clean-up the platform locking that may occur with an EventToCommand implementation, it would not support Mixed-Mode Bidirectional implementations (though this can be solved using a Service). 

I have personally implemented something like this before so I am familiar with the concept, whether or not you feel it has any value to be added to the Framework in a vNext release I'll leave up to you.

### PR DESCRIPTION
**Pull request by vaecors**
Implements an EventArgsConverterBase class and interface that can be used for translating EventArgs objects into neutral objects.

Closes #1308 

### Checklist

- [x] I have included examples or tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file
- [x] I have cleaned up the commit history (use rebase and squash)

### Changes proposed in this pull request:

This PR creates the dependency property EventArgsConverter for EventToCommand as discussed in #1308. Below is a basic example that converts a PreviewMouseDoubleClick event against a ListBoxItem into a PersonModel tunneled to a Command.

**Converter:**
```csharp
public class PersonListEventArgsConverter : EventArgsConverterBase<MouseEventArgs>
{
	protected override object Convert(object sender, MouseEventArgs args)
	{
		if ( !( args.OriginalSource is DependencyObject dependencyObject ) )
		{
			return null;
		}

		ListBoxItem item = ParentOfType<ListBoxItem>(dependencyObject);

		if ( !( item.DataContext is PersonModel model ) )
		{
			return null;
		}

		return model;
	}

	private static T ParentOfType<T>(DependencyObject dependencyObject)
		where T : DependencyObject
	{
		DependencyObject parent = VisualTreeHelper.GetParent(dependencyObject);
		if ( parent == null )
		{
			return default;
		}

		T result = parent as T;
		return result ?? ParentOfType<T>(parent);
	}
}
```

**Model:**
```csharp
public class PersonModel : ModelBase
{
	private static int _nextId;

	public PersonModel()
	{
		Id = Interlocked.Increment(ref _nextId);
	}

	public string FirstName { get; set; }

	public int Id { get; set; }

	public string LastName { get; set; }
}
```

**ViewModel:**
```csharp
public class MainViewModel : ViewModelBase
{
	private readonly IMessageService _messageService;

	public MainViewModel(IMessageService messageService)
	{
		Argument.IsNotNull(() => messageService);

		_messageService = messageService;

		ShowItemAsyncCommand = new TaskCommand<PersonModel>(OnShowItemAsync);
	}

	public FastObservableCollection<PersonModel> Items { get; set; } = new FastObservableCollection<PersonModel>();

	public TaskCommand<PersonModel> ShowItemAsyncCommand { get; }

	protected override Task InitializeAsync()
	{
		using ( Items.SuspendChangeNotifications() )
		{
			Items.Add(new PersonModel {FirstName = "John", LastName = "Doe"});
			Items.Add(new PersonModel {FirstName = "Jane", LastName = "Doe"});
		}

		return base.InitializeAsync();
	}

	private async Task OnShowItemAsync(PersonModel model)
	{
		await _messageService.ShowInformationAsync($"Id: {model.Id} - Name: {model.FirstName} {model.LastName}");
	}
}
```

**View:**
```xaml
<catel:Window x:Class="WpfApp1.Views.MainWindow"
			  xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
			  xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
			  xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
			  xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
			  xmlns:catel="http://schemas.catelproject.com"
			  xmlns:i="http://schemas.microsoft.com/expression/2010/interactivity"
			  xmlns:converters="clr-namespace:WpfApp1.Converters"
			  mc:Ignorable="d"
			  Title="MainWindow" Height="450" Width="800">

	<catel:Window.Resources>
		<converters:PersonListEventArgsConverter x:Key="PersonListEventArgsConverter" />
	</catel:Window.Resources>
	<ListBox ItemsSource="{Binding Items}">
		<i:Interaction.Triggers>
			<i:EventTrigger EventName="PreviewMouseDoubleClick">
				<catel:EventToCommand Command="{Binding ShowItemAsyncCommand}" PassEventArgsToCommand="True" EventArgsConverter="{StaticResource PersonListEventArgsConverter}" />
			</i:EventTrigger>
		</i:Interaction.Triggers>
		<ListBox.ItemTemplate>
			<DataTemplate>
				<VirtualizingStackPanel Orientation="Horizontal">
					<TextBlock Text="{Binding FirstName}" />
					<TextBlock Text="{Binding LastName}" Margin="5 0 0 0" />
				</VirtualizingStackPanel>
			</DataTemplate>
		</ListBox.ItemTemplate>
	</ListBox>
</catel:Window>
```

### CODE REVIEW

**Comment by GeertvanHorrik** [comment]
Fantastic documentation, will check the code now.

File: doc/history.txt [hunk 1]
```diff
@@ -43,6 +43,7 @@ Classes / members marked as obsolete
 Added/fixed
 ===========
 (x) #1302 Implement support for disposing of IDisposable ViewModels in WindowLogic
+(+) #1308 Add EventArgsConverter Support in EventToCommand
```
Comment (GeertvanHorrik) [inline doc/history.txt]: Although this is an unwritten rule, new features go on top, then changes, then bug fixes, then removals. No need to change this for the PR, but I do want you to be aware for a potential next time :-)

**Comment by GeertvanHorrik** [comment]
The only things that are missing are unit tests, but this looks really solid, well done!

File: doc/history.txt (hunk 1 above)
Comment (vaecors) [reply doc/history.txt]: Whoops, that's what happens when I do that one last.

**Pull request merged**